### PR TITLE
[WFCORE-3666] Provide Elytron alternative to RoleMappingLoginModule

### PR DESCRIPTION
**Pull request by nekdozjam**
https://issues.jboss.org/browse/WFCORE-3666
https://issues.jboss.org/browse/EAP7-1008

### CODE REVIEW

File: elytron/WFCORE-3666-alternative-rolemapping.adoc [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+=== Hard Requirements
+
+* Implement a Map backed RoleMapper for a -> b mapping. There should be options to keep/remove the original mapped role, and to keep/remove roles that weren't mapped. The map would map a role to one or more new roles.
+[source,xml]
+----
+<mappers>
+...
+	<mapped-role-mapper name="my-mapped-role-mapper" removeMapped="true" removeNotMapped="false" />
+		<role-mapping from="roleToMap" to="mappedRole1 mappedRole2" />
```
Comment (mchoma) [inline elytron/WFCORE-3666-alternative-rolemapping.adoc]: In model this is STRING and items are delimited with comma?
What about using LIST of STRING in model?

File: elytron/WFCORE-3666-alternative-rolemapping.adoc [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+		<role-mapping from="foo" to="joe bar" />
+	</mapped-role-mapper>
+...
+</mappers>
+
+----
+
+=== Nice-to-Have Requirements
```
Comment (mchoma) [inline elytron/WFCORE-3666-alternative-rolemapping.adoc]: "Nice-to-have" that mean it is not in scope of this RFE, right?

File: elytron/WFCORE-3666-alternative-rolemapping.adoc [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+...
+</mappers>
+
+----
+
+* By combining this regex based mapper and mapped role mapper we could achieve similar functionality as mapped-regex-realm-mapper, but for roles.
+
+=== Non-Requirements
+
+//== Implementation Plan
```
Comment (mchoma) [inline elytron/WFCORE-3666-alternative-rolemapping.adoc]: Can be removed?

File: elytron/WFCORE-3666-alternative-rolemapping.adoc [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+* https://issues.jboss.org/browse/WFCORE-3666[WFCORE-3666]
+
+=== Related Issues
+
+* https://issues.jboss.org/browse/ELY-1591[ELY-1591]
+* https://issues.jboss.org/browse/EAP7-1008[EAP7-1008]
+
+=== Dev Contacts
+
+* mailto:{email}[{author}]
```
Comment (mchoma) [inline elytron/WFCORE-3666-alternative-rolemapping.adoc]: please fill

File: elytron/WFCORE-3666-alternative-rolemapping.adoc [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+=== QE Contacts
+
+* TBD
```
Comment (mchoma) [inline elytron/WFCORE-3666-alternative-rolemapping.adoc]: you can fill mchoma@redhat.com

File: elytron/WFCORE-3666-alternative-rolemapping.adoc [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+//== Implementation Plan
+////
+Delete if not needed. The intent is if you have a complex feature which can 
+not be delivered all in one go to suggest the strategy. If your feature falls 
+into this category, please mention the Release Coordinators on the pull 
+request so they are aware.
+////
+== Test Plan
+
+* Mapping functionality tests will be added to Elytron.
```
Comment (mchoma) [inline elytron/WFCORE-3666-alternative-rolemapping.adoc]: And here you mean some unit tests in Elytron TS

File: elytron/WFCORE-3666-alternative-rolemapping.adoc (hunk 2 above)
Comment (nekdozjam) [reply elytron/WFCORE-3666-alternative-rolemapping.adoc]: Yes

File: elytron/WFCORE-3666-alternative-rolemapping.adoc (hunk 5 above)
Comment (nekdozjam) [reply elytron/WFCORE-3666-alternative-rolemapping.adoc]: Will do

File: elytron/WFCORE-3666-alternative-rolemapping.adoc (hunk 4 above)
Comment (nekdozjam) [reply elytron/WFCORE-3666-alternative-rolemapping.adoc]: Will do, although in my case this works as intended

File: elytron/WFCORE-3666-alternative-rolemapping.adoc (hunk 1 above)
Comment (nekdozjam) [reply elytron/WFCORE-3666-alternative-rolemapping.adoc]: Model is LIST of STRING. And xml I was thinking STRING_LIST parser.

File: elytron/WFCORE-3666-alternative-rolemapping.adoc [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+
+=== Hard Requirements
+
+* Implement a Map backed RoleMapper for a -> b mapping. There should be options to keep/remove the original mapped role, and to keep/remove roles that weren't mapped. The map would map a role to one or more new roles.
+* The format of role to be mapped is simple string, roles to which the role should be mapped are comma delimited in a string. In model this makes up a MAP of LIST, where the keys represent roles to be mapped and the representing values are LISTs of STRINGs, which represent the resulting roles.
+[source,xml]
+----
+<mappers>
+...
+	<mapped-role-mapper name="my-mapped-role-mapper" keep-mapped="true" keep-non-mapped="false" />
```
Comment (mchoma) [inline elytron/WFCORE-3666-alternative-rolemapping.adoc]: @aloubyansky can you confirm this is proper named list, which will fit to galleon provisioning. And is not same case as in WFCORE-3596?

Comment (aloubyansky) [reply elytron/WFCORE-3666-alternative-rolemapping.adoc]: It is not the same in a sense that the items here are identifiable (by keys in the map) and the list here being actually a set, afaiu. Unfortunately we don't have SET as a built-in type. So, I'd say it's ok. Thanks.

**Comment by kabir** [comment]
@nekdozjam Can you please add in the new Community Documentation section I just mentioned on wildfly-dev?

**Pull request merged**